OpenRA.Game/Server/OrderBuffer.cs

```diff
@@ -0,0 +1,82 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Server
+{
+	/*
+	 * OpenRA orders do not need to be frame-specific, as long as each client resolves the same orders in the same order
+	 * they should be able to maintain a deterministic simulation with a mostly expected result.
+	 *
+	 * This allows the server to act as a dumb "master order synchronizer", simply acting as the authority and central
+	 * point of synchronization on which orders to resolve each frame.
+	 *
+	 * The buffer is fairly simple, it maintains a list of all incoming order packets, minus their associated frames,
+	 * so that all orders can be batched for a single frame to all clients at a time of the server's choosing.
+	 *
+	 * A mechanism for clients to inform/request slowdown should be provided so that clients which are overloaded do
+	 * not fall behind indefinitely.
+	 */
+	public class OrderBuffer
+	{
+		// Maintaining a queue of the serialized order lists prevents us from having to deserialize the orders
+		// or copy them on receive. They should be able to be written the TCP stream without being copied.
+		// The count of the byte[] queue is used as the number of frames to ack.
+		// Since the connection is reliable, clients will understand what the ack count means.
+		readonly Dictionary<int, List<byte[]>> clientOrdersBuffer = new Dictionary<int, List<byte[]>>();
```

Comment:
Couldn't this be a list of memory streams already?

Reply:
We can defer that to a follow-up.